### ISSUE TITLE
Migration concepts / best practices

### ISSUE THREAD

**Issue by xavierLowmiller** [vapor/fluent#658]
I'm attempting a migration that is more than simply adding/removing fields.

Consider the following object:
```swift
final class User: PostgreSQLModel {
  var id: Int? = nil
  var mailAddress: String
  var isAdmin: Bool
}
```

I'm trying to migrate to this structure:
```swift
final class User: PostgreSQLModel {
  enum Role { case tester, developer, admin }
  var id: Int? = nil
  var mailAddress: String
  var role: Role
}
```

My current solution involves 3 steps:
1. Add the new field `role` from the database
2. Migrate the data
3. Remove the field `isAdmin` from the database

This works for a database that was created using the old structure, but fails on new databases, because the field `isAdmin` is never created and can therefore never be migrated or removed.
Also, the field `role` would be added twice, causing another error.

I'm not looking for a solution for my particular use case. Rather, I'm trying to understand how migrations are supposed to work in mutating cases like this: Do I 
1. Always keep the initial fields and all migrations around
2. Run the migrations and change the models to the desired shapes after

**Comment by 0xTim** [comment]
@xavierLowmiller (I'm assuming you're referring to Fluent 3 here and not Fluent 4 which makes things a lot easier)

So you've hit on one of the problems of migrations and a fundamental issue with Fluent 3. Because of a type safety in dealing with ever-changing migrations in a stringly typed database you're fairly constrained., which has been discussed in vapor/fluent-kit#139. Additionally, Fluent records which migrations it has run and will never run them again. But there are some rules you should definitely follow:

1. Implement your migrations manually - always conform to `Migration` yourself and implement `prepare(on:)` yourself, _without_ using `addProperties(to:)`. That will catch you out. 
2. Each migration is a separate type, you should never change the original migration. So in your case you'll have two migrations. The first for `User` which creates the database and all the original fields. You then have a second migration which converts the `isAdmin` field to a `role` enum. 

Basically, you want a record of what has changed in each migration, which allows any database, no matter what migration stage it's in to get to the correct state. So that's your existing prod database that's had everything, a database for a branch that has run 2 out of 3 migrations or a fresh database used for test. Now in your case, because you're deleting an original field, things get a bit complicated. You have a couple of options - you can either implement as much as your original migration as possible with Fluent (create the table, add `id` and `mailAddress`) and then chain the execution of that command (in the same migration) to some raw SQL to create the original `isAdmin` field. Or you. can do it all in one DB query with raw SQL. Hope that helps!

**Comment by xavierLowmiller** [comment]
Thank you very much, this helps me out tremendously!

Yes, I was talking about Fluent 3, and am looking forward to Fluent 4.

My solution is now to define a model that has exactly the fields as my original `User` model just for the migration, which has the same `entity`:


```swift
extension User: Migration {
    private final class OriginalUser: PostgreSQLModel {
        static let entity = "User"
        var id: Int?
        var mailAddress: String
        var isAdmin: Bool
    }

    static func prepare(on conn: PostgreSQLConnection) -> Future<Void> {
        Database.create(OriginalUser.self, on: conn) { builder in
            try OriginalUser.addProperties(to: builder)
        }
    }

    static func revert(on conn: PostgreSQLConnection) -> EventLoopFuture<Void> {
        Database.delete(OriginalUser.self, on: conn)
    }
}
```

Having played with the Vapor 4 beta, I think this pattern of keeping all the migrations around will be far more apparent 👍.
